New cookie banner and consent message

riknoll opened the pull request:
__Note: Please review this PR carefully! Especially check the events I have labelled as constituting interactive consent to make sure they are the result of explicit user actions. If you have questions about something, message me.__

This PR enables the application insights cookies along with a new UI for the cookie consent message. A localized banner now appears on top of everything else and disappears automatically when the user interacts with the page. To see the banner in action on another site, visit https://www.microsoft.com/fr-fr in a private-browsing window. When the banner disappears, a cookie is stored indicating that the user consents to cookies and the _next_ time they visit application insights will be allowed to drop cookies (but not in the current session, AI doesn't really allow conditional initialization like that).

As part of this change I also added `docfiles/pxtweb` for holding onto any code that belongs on _every_ page served on the site. If you add any code there, make sure it doesn't have any dependencies. I also removed the pxtarget flag for disabling the cookie message; targets can override it by including their own tracking.html (but they shouldn't).

Known issues:
* In IE, the cookie banner appears over the header in makecode.com (rather than pushing it down). On small screens it can cover up the entire header

Once this is merged it also needs to be ported to V0 and V2.

riknoll commented:
BTW, the language cookie is unaffected by this.

pelikhan commented on docfiles/docs.js:
```diff
@@ -50,14 +50,14 @@ function dirAuto($el) {
 
 function searchSubmit(form) {
-    if (pxt && pxt.tickEvent) pxt.tickEvent("docs.search", { 'source': form.id })
+    if (pxt && pxt.tickEvent) pxt.tickEvent("docs.search", { 'source': form.id }, /*interactiveConsent*/true)
```
this notation is going to hard to enfore in the future. Instead, you could define an Options interface 
```
interface Options {
    interactiveConsent?: boolean;
}
```
and do
```
tickEvent(...., { interactiveConsent: true })
```
that way, TypeScript forces us to explain the meaning of that boolean value.

riknoll replied:
Sure

pelikhan commented on docfiles/pxtweb/cookieCompliance.ts:
```diff
@@ -0,0 +1,217 @@
+                    return;
+                }
+                catch (e) {
+                    cb(new Error("Bad response from server: " + resp.body))
```
bail out and consider that cookies are not permitted. Maybe the user is behind a hotel front page and the banner returns HTML garbage.

riknoll replied:
That's what I'm doing already. This will pass an error to the callback which is handled above (see `bannerErr` check)

pelikhan commented on docfiles/pxtweb/cookieCompliance.ts:
```diff
@@ -0,0 +1,217 @@
+                    return;
+                }
+            }
+            cb(new Error("didn't get 200 response: " + resp.status + " " + resp.body));
```
Handle failure gracefully.

riknoll replied:
What's un-graceful about this? I'm passing an error, not throwing it.

pelikhan commented on docfiles/pxtweb/cookieCompliance.ts:
```diff
@@ -0,0 +1,217 @@
+                info.Css.forEach(injectStylesheet)
+            }
+
+            all(info.Js || [], injectScriptAsync, msccError => {
```
it practices there's only 1 javascript file (and there should never be more than one)

riknoll replied:
The API returns an array so I support an array.

riknoll commented:
Answers:

1. __Is this tested in all browsers?__ Tested Chrome, Edge, IE, and Firefox on Windows
2. __What happens when the user goes offline?__ The request to the service will fail and app insights will load without cookies. If the cookies are already present, they will not be cleared (but may be the next time the page is loaded).
3. __What happens if the mscc.js file fails to load?__ App insights will be loaded without cookies. I think the banner might also be displayed; I have to check that.
4. __Is this happening in the background or blocking the editor load?__ Everything is happening in the callback to the `XMLHttpRequest`, so it should be async.

pelikhan commented:
LGTM

riknoll commented:
I just found a bug that I'm fixing; I'll comment when I'm finished and do the merge.

abchatra commented:
Lets port to v0 once merged and tested.

abchatra commented:
I don't think this needs to be ported to v2.

riknoll commented:
@pelikhan @abchatra please take a look at the last commit; I added queueing of events to prevent races as the editor loads and fixed a bug that caused analytics not to load in the editor.

pelikhan commented on docfiles/pxtweb/cookieCompliance.ts:
```diff
@@ -126,6 +138,10 @@ namespace pxt {
             loadAI(includeCookie)
+            analyticsLoaded = true;
+            while (pendingCallbacks.length) {
+                pendingCallbacks.pop()();
```
are you supposed to call these callbacks?

riknoll replied:
I am calling them. Maybe I should add a comment since I had to stare at that line for a bit to figure that out...

pelikhan replied:
Please. I'm really confuse about what's happening here.

riknoll replied:
There are two sets of parentheses, one for `pop` and one to call the callback. I'll add a comment.

pelikhan replied:
what happens when the client overrides pxtError / pxtEvent?

riknoll replied:
Not sure what you mean by that

pelikhan commented:
@riknoll why aren't we simply storing the data instead of all the callbacks?

pelikhan commented on docfiles/pxtweb/cookieCompliance.ts:
```diff
@@ -0,0 +1,233 @@
+        }
+    }
+
+    function httpGetAsync(url: string, cb: Callback<HttpResponse>) {
```
don't we have helpers to do this already?

pelikhan commented on pxtlib/analytics.ts:
```diff
@@ -1,13 +1,60 @@
+
+        const rexp = pxt.reportException;
+        const rexpHandled = wrap((ev: ExceptionData) => reportExceptionCore(ev.err, ev.data));
+        pxt.reportException = function (err: any, data: pxt.Map<string>): void {
```
this function can be overriden, does it matter? it won't be using wrap most likely.

riknoll replied:
I don't think it matters. Who overrides this BTW? The only place I could find in the codebase is the editor controller business and that shouldn't matter.

riknoll replied:
If someone overrides it and they don't call the original then it's a moot point anyway.

riknoll commented:
I am storing the data (see the `wrap` function). The reason I also have to do the `pendingCallbacks` stuff is because the analytics code lives in pxtlib and AI gets loaded by pxtweb and I'm trying not to create any hard dependencies between the two. If you want, I could instead move the implementations of `tickEvent`, `reportError`, etc. into `pxtweb` and just leave the declarations in `pxtlib`.

pelikhan commented:
Sounds simpler.

________________________________
From: Richard Knoll <notifications@github.com>
Sent: Wednesday, January 31, 2018 2:59:39 PM
To: Microsoft/pxt
Cc: Peli de Halleux; Mention
Subject: Re: [Microsoft/pxt] New cookie banner and consent message (#3767)


I am storing the data (see the wrap function). The reason I also have to do the pendingCallbacks stuff is because the analytics code lives in pxtlib and AI gets loaded by pxtweb and I'm trying not to create any hard dependencies between the two. If you want, I could instead move the implementations of tickEvent, reportError, etc. into pxtweb and just leave the declarations in pxtlib.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FMicrosoft%2Fpxt%2Fpull%2F3767%23issuecomment-362099917&data=02%7C01%7Cjhalleux%40microsoft.com%7Ccfd1543602b544da51b808d568fe4f25%7Cee3303d7fb734b0c8589bcd847f1c277%7C1%7C0%7C636530363850635984&sdata=z4kaa6gCDoD8iO3Et1564KhJOtRNAcGj8vjChrMuyBE%3D&reserved=0>, or mute the thread<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAD-4KXrfMraRonDyLKwc_bRzfhItnuk4ks5tQPBbgaJpZM4RtnMW&data=02%7C01%7Cjhalleux%40microsoft.com%7Ccfd1543602b544da51b808d568fe4f25%7Cee3303d7fb734b0c8589bcd847f1c277%7C1%7C0%7C636530363850635984&sdata=qkW%2Fn8RfjZwx6p%2B6ISP2xOrTtIjM6z3FmWTEXM7QBls%3D&reserved=0>.

riknoll commented:
Will do

riknoll commented:
@pelikhan check it out. Also, I'm removing the DO NOT MERGE so feel free to merge this if you're happy with it. The backend is already set up.

pelikhan commented on docfiles/pxtweb/cookieCompliance.ts:
```diff
@@ -71,13 +102,18 @@ namespace pxt {
-            cb();
+    export function aiTrackEvent(id: string, data?: any, measures?: any) {
+        if (!eventLogger) {
+            eventLogger = new TelemetryQueue((a, b, c) => (window as any).appInsights.trackEvent(a, b, c));
```
is appInsights always defined here?

pelikhan commented on docfiles/pxtweb/cookieCompliance.ts:
```diff
@@ -71,13 +102,18 @@ namespace pxt {
+
+    export function aiTrackException(err: any, kind?: string, props?: any) {
+        if (!exceptionLogger) {
+            exceptionLogger = new TelemetryQueue((a, b, c) => (window as any).appInsights.trackException(a, b, c));
```
is appInsights always defined here?

pelikhan commented:
What happens in statically packaged editors?

pelikhan commented:
Tested in a number of places and it looks good.
